pkg/app/piped/imagewatcher/watcher.go

```diff
@@ -156,51 +155,63 @@ func (w *watcher) determineUpdates(ctx context.Context, repoID string, repo git.
 	if !ok {
 		return nil, fmt.Errorf("configuration file for Image Watcher not found: %w", err)
 	}
-
-	updates := make([]config.ImageWatcherTarget, 0)
-	for _, target := range cfg.Targets {
-		if provider.Name() != target.Provider {
-			continue
-		}
-		outdated, err := checkOutdated(ctx, target, repo, provider)
-		if err != nil {
-			return nil, fmt.Errorf("failed to check the image is outdated: %w", err)
-		}
-		if outdated {
-			updates = append(updates, target)
-		}
-	}
-	return updates, nil
+	return cfg, nil
 }
 
-// checkOutdated checks if the image defined in the given target is identical to the one in image provider.
-func checkOutdated(ctx context.Context, target config.ImageWatcherTarget, repo git.Repo, provider imageprovider.Provider) (bool, error) {
+// updateOutdatedImage first compares the image in git repository and one in image provider.
+// Then pushes rewritten one to the git repository if any deviation exists.
+func (w *watcher) updateOutdatedImage(ctx context.Context, target *config.ImageWatcherTarget, repo git.Repo, provider imageprovider.Provider) error {
+	// Fetch from the image provider.
 	i, err := provider.ParseImage(target.Image)
 	if err != nil {
-		return false, err
+		return fmt.Errorf("failed to parse image string \"%s\": %w", target.Image, err)
 	}
 	// TODO: Control not to reach the rate limit
-	imageRef, err := provider.GetLatestImage(ctx, i)
+	imageInRegistry, err := provider.GetLatestImage(ctx, i)
 	if err != nil {
-		return false, err
+		return fmt.Errorf("failed to get latest image from %s: %w", provider.Name(), err)
 	}
 
-	yml, err := ioutil.ReadFile(filepath.Join(repo.GetPath(), target.FilePath))
+	// Fetch from the git repository.
+	path := filepath.Join(repo.GetPath(), target.FilePath)
+	yml, err := ioutil.ReadFile(path)
 	if err != nil {
-		return false, err
+		return fmt.Errorf("failed to read file: %w", err)
 	}
 	value, err := yamlprocessor.GetValue(yml, target.Field)
 	if err != nil {
-		return false, err
+		return fmt.Errorf("failed to get value at %s in %s: %w", target.Field, target.FilePath, err)
 	}
-	v, ok := value.(string)
+	imageInGit, ok := value.(string)
 	if !ok {
-		return false, fmt.Errorf("unknown value is defined at %s in %s", target.FilePath, target.Field)
+		return fmt.Errorf("unknown value is defined at %s in %s", target.FilePath, target.Field)
+	}
+
+	outdated := imageInRegistry.String() != imageInGit
+	if !outdated {
+		return nil
 	}
-	return imageRef.String() != v, nil
-}
 
-func update(targets []config.ImageWatcherTarget) error {
-	// TODO: Make it possible to push outdated images to Git
+	// Update the outdated image.
+	newYml, err := yamlprocessor.ReplaceValue(yml, target.Field, imageInRegistry.String())
+	if err != nil {
+		return fmt.Errorf("failed to replace value at %s with %s: %w", target.Field, imageInRegistry, err)
+	}
+	w.mu.Lock()
+	f, err := os.OpenFile(path, os.O_WRONLY|os.O_TRUNC, os.ModeAppend)
```

Comment:
Writing directly into the `git.Repo` may cause unexpected behaviors.
- other goroutines are also accessing this local directory too
- adding a new commit may cause a pull error in the future

So I think we have to clone it into a temporary directory for pushing the changes.

Reply:
Sounds great idea, will do so!

Reply:
This change will be unneeded if we change to spawn goroutines for each repository according to your suggestion in the above comment, right?

Reply:
Damm still remains an issue adding a new commit may cause a pull error in the future.

Reply:
Leave TODO comments so may I address them when I work on the issue to spawn goroutines for each repo?

Reply:
Sure.